### ISSUE TITLE
Hammerspoon behaving erratically wrt Accessibility

### ISSUE THREAD

**Issue by chdiza** [Hammerspoon/hammerspoon#210]
I confess that I'm having a really hard time getting Accessibility to STAY enabled for Hammerspoon.  Slate, Mjolnir, and Hydra were all such that once accessibility was granted, it stayed granted (until one replaced the app with a newer version or different build).

For me, on multiple machines, Hammerspoon is losing its already-granted Accessibility in seemingly random ways.  

For example, sometimes I see the console pop up with the "nil value" nonsense that indicates an Accessibility problem (since I know my init doesn't have errors, as it loads and runs other times); I then go grant Access, and then Hammerspoon works again; but then (say) ten minutes later, the very same open instance of Hammerspoon that was just working will throw the console up with the "nil value" business.

I don't know what kind of info I can provide to help solve this.  Please let me know if I can help make this issue solvable.  I'm on the verge of giving up and going back to Hydra or Slate.

**Comment by cmsj** [comment]
Odd, I've never seen it get dropped from a running instance.
Have you been running your own builds on these machines? The system can become confused if it has multiple instances of an accessibility app where some instances are signed and some aren't.

http://www.hammerspoon.org/faq/ has a link to a way to repair the confusion

**Comment by cmsj** [comment]
(I'm gonna close this out, because I very strongly suspect that the issue is OS confusion, but please re-open it if you're not able to resolve things with that link)

**Comment by chdiza** [comment]
I have been running my own builds.

The link that the FAQ pointed to says to not keep multiple copies of an app on one's machine.  I didn't know that could mess with Accessibility, but with other apps the existence of multiple copies has caused occasional weirdness.  (This latter experience is what led me to make the earlier PR about not leaving an extra Hammerspoon.app lying around.)

I have been manually destroying `~/Library/Developer` after every time I build Hammerspoon, so there are no multiple copies of Hammerspoon.app anymore.  (I guess technically a dupe exists very briefly before I destroy the folder).

I'll try the DB trick in the link.

**Comment by cmsj** [comment]
Something I've been meaning to experiment with, is having Spotlight ignore the build directories. Not sure if it would help, but it might!

**Comment by asmagill** [comment]
Actually, it's probably worth doing if it's not too difficult... I've run into issues before when starting Hammerspoon via "open -a Hammerspoon" where I don't always get the version I expect... since I use this approach to allow for a complete restart of the app when testing new code (https://github.com/asmagill/hammerspoon-config <https://github.com/asmagill/hammerspoon-config>, if anyone is interested), I've taken to looking up the actual app's full path via one of the extra modules I've written and invoking that specifically to ensure I stay with the same version of the core app. (https://github.com/asmagill/hammerspoon_asm <https://github.com/asmagill/hammerspoon_asm>, again if interested, specifically the _paths array of extras.)


I never bothered to dig deeper to figure out what search arrangement OS X uses, as the search path seems to change sometimes for reasons I don't understand, but Spotlight seems likely.

**Comment by chdiza** [comment]
While it might be a good idea to have Spotlight ignore some stuff, I don't see how it can be behind my particular problem.  The reason is that I have Spotlight disabled.  It neither searches nor indexes.  And it's been disabled since long before I ever cloned Hammerspoon's repo or downloaded a prebuilt copy.

**Comment by cmsj** [comment]
@chdiza aha, fair enough. I guess its Gatekeeper that keeps track. Did clearing out the database entries help?

**Comment by chdiza** [comment]
> Did clearing out the database entries help?

Unclear, but I suspect not.  Reason: Quitting Hspoon, then `defaults delete`-ing Hammerspoon's preference plist file in ~/library/preferences, then restarting Hammerspoon resulted in the following infelicity.  The preferences GUI window popped up (like it should for a "fresh" run), but it said Accessibility was NOT enabled and showed a yellow circle.  But of course, Accessibility WAS enabled; it was enabled before I quit Hspoon and I didn't touch it or install a new Hspoon in the meantime.  Also, I could use the running Hspoon (the one whose GUI pref window was open) without springing the Console.

So I think something is wrong with Hspoon's ability to accurately detect whether Accessibility is enabled.  This may or may not be causally bound up with the main complaint in this thread, but it seemed potentially relevant.  It leads me to believe that the problem doesn't like *solely* with a confused DB file.  It may or may not actually be Hspoon's fault that it can't always do accurate detection.

**Comment by cmsj** [comment]
@chdiza fwiw, in my experience, getting the yellow circle in Hammerspoon, but seeing a tick in System Preferences, means the TCC database has gotten confused. Could you reply with the output of:

sudo sqlite3 /Library/Application\ Support/com.apple.TCC/TCC.db 'select * from access where client like "%hammerspoon%"'

I'm going to re-open this issue as there may be a deeper mystery at play here.

**Comment by chdiza** [comment]
I'll post that output later when I'm at a Hammerspoony machine.

> getting the yellow circle in Hammerspoon, but seeing a tick in System Preferences, means the TCC database has gotten confused

If that's true in this particular case, then the database got confused as a result of the mere act of running and then quitting Hammerspoon, followed by a deletion of the prefs file.

**Comment by chdiza** [comment]
> Could you reply with the output of:
> sudo sqlite3 /Library/Application\ Support/com.apple.TCC/TCC.db 'select * from access where client like "%hammerspoon%"'


I get:
````
kTCCServiceAccessibility|org.hammerspoon.Hammerspoon|0|1|0|��
````

I dunno what's up with the rightmost two chars; they come out that way on both iTerm and Terminal.app with UTF-8 as the encoding.

**Comment by cmsj** [comment]
@chdiza ok, that output looks fine, so it's not that the signing vs accessibility stuff is confused. Curiouser and curiouser :(

**Comment by 0942v8653** [comment]
Those characters on the right are the `csreq` blob, which looks related to code signing. It might change every update (invalidating the new app bundle; System Preferences unchecking and rechecking would remove it from the database and add it again, getting a new one).

**Comment by chdiza** [comment]
Now I'm noticing that Hspoon loses its Accessibility status after random reboots of the machine.

**Comment by cmsj** [comment]
@chdiza did this settle down? Or is it still being confusing and weird?

**Comment by chdiza** [comment]
It's partly settled.

It seems to "lose" the exalted Acessible status if I quit it, then start Slate and enable its accessibility, then quit Slate and restart HS.

The output of the suggested SQL command above remains the same.

**Comment by cmsj** [comment]
I've looked into this issue several times this year and I'm fairly well convinced that we're doing everything correctly. We've also not had any other reports of this, so if it's still happening, I'm rather tempted to suggest that it's an oddity with your particular system, unfortunately.

If anyone disagrees, please re-open this :)
